### PR TITLE
feat: update/settings connection status

### LINKED ISSUE #711: Rewrite Settings Screen to use peerConnectionStatus vuex state for displayed connection state.
**Is your feature request related to a problem? Please describe.**
Currently, this [line of code](https://github.com/ambianic/ambianic-ui/blob/master/src/views/Settings.vue#L22) within the Settings page has an `if` condition that evaluates to true when there is an `edgePeerId`.  This should be corrected to use the `peerConnectionStatus`.

**Describe the solution you'd like**
This [line of code](https://github.com/ambianic/ambianic-ui/blob/master/src/views/Settings.vue#L22) should be rewritten to use the `peerConnectionStatus` to sync the connection state with the connection icon in the Navbar.

This feature request is a follow-up to this [pull request](https://github.com/ambianic/ambianic-ui/pull/708).
Cc: @ivelin

### PR DESCRIPTION
**Pull request by vickywane**
## Purpose
this pull request will fix [this issue](https://github.com/ambianic/ambianic-ui/issues/711)

## Approach
This pull request adjusts the `Settings.vue` component to better handle `PEER_DISCONNECTED` connection status, and also updates the connection status icon.

When the PWA moves into the `PEER_DISCONNECTED` state, the connection state text is updated accordingly, and the progress element and release version element are immediately hidden. The attached snapshot below shows the modifications made;

Screenshot_20210902_011243

## Merge Checklist
- [x]  The code change is tested and works locally.
- [x]  The user and dev documentation is up to date.
- [x]  There is no commented out code in this PR.
- [x]  No lint errors (use flake8)
- [x]  100% test coverage

### CODE REVIEW

**Comment by vickywane** [comment]
@ivelin 

Further questions for clarification while this PR is being worked upon;
- Are there elements such as Peer Name and ID that should be further removed when the PWA is in a `PEER_DISCONNECTED` state?

**Comment by vickywane** [comment]
> This question is a bit vague. Not sure how it relates to the edge status panel in the settings page.

Alright

**Comment by ivelin** [comment]
@vickywane could you please resolve the merge conflicts and resubmit?

File: src/views/Settings.vue [hunk 1]
```diff
@@ -349,44 +355,46 @@ export default {
...
+      isEdgeConnected: state =>
+        state.pnp.peerConnectionStatus === PEER_CONNECTED,
+      pnp: state => state.pnp,
+      edgePeerId: state => state.pnp.remotePeerId,
+      peerFetch: state => state.pnp.peerFetch,
+      edgeVersion: state => state.edgeDevice.edgeSoftwareVersion
+    })
+  },
   watch: {
     edgeAddress (value) {
```
Comment (ivelin) [inline src/views/Settings.vue:399]: @vickywane Does this one make sense to you? Why would we update the edgeVersion when the remoteEdgeAddress field changes? Instead, we should update the edgeDetails info AFTER a new peer connection is established.

File: src/views/Settings.vue [hunk 2]
```diff
@@ -396,6 +404,9 @@ export default {
       if (isConnected && !this.edgeVersion) {
```
Comment (ivelin) [inline src/views/Settings.vue:404]: @vickywane Sine edgeVersion is synced with the vuex store, I think we should fetch edgeversion info each time a new connection is established. The reason is that the edge device version might have changed due to an automatic edge software update. The UI needs to be able to pick that up. Please think about it and cover with test cases.

File: src/views/Settings.vue [hunk 3]
```diff
@@ -303,11 +305,15 @@ export default {
     return {
       edgeAddress: undefined,
```
Comment (ivelin) [inline src/views/Settings.vue:306]: @vickywane this variable name confuses me each time I review this code. Let's rename it to something more self explanatory. Maybe `manualEdgeAddress` so it is clear this is only used when the user enters manually a remote edge address value and it has nothing to do with the currently connected edge address or the auto discovery mode edge address.

File: src/views/Settings.vue [hunk 4]
```diff
@@ -349,44 +355,46 @@ export default {
...
-      let step = 1
+    handleConnectionStep () {
       switch (this.peerConnectionStatus) {
         case PEER_DISCONNECTED:
+          this.peerConnectionIcon = 'cloud-off-outline'
+          this.peerConnectionText = PEER_DISCONNECTED_TEXT
+          break
         case PEER_CONNECTION_ERROR:
-          step = 1
+          this.connectStep = 1
```
Comment (ivelin) [inline src/views/Settings.vue:366]: @vickywane thanks for making these variable names more self explanatory. Helps with reviews.

**Comment by ivelin** [comment]
@vickywane is there a way to solve these repetitive annoying warning messages for 
`Variable 'attrs' is already declared in the upper scope` in `ListItem.vue` ?

**Review by ivelin — CHANGES_REQUESTED** [review]

**Comment by vickywane** [comment]
> @vickywane is there a way to solve these repetitive annoying warning messages for Variable 'attrs' is already declared in the upper scope in ListItem.vue ?

I have fixed this error

File: src/views/Settings.vue (hunk 4 above)
Comment (vickywane) [reply src/views/Settings.vue]: Alright

File: src/views/Settings.vue (hunk 3 above)
Comment (vickywane) [reply src/views/Settings.vue]: Alright. 

I have modified this state and updated other references

File: src/views/Settings.vue (hunk 2 above)
Comment (vickywane) [reply src/views/Settings.vue]: I have addressed this concern by removing the second `!this.edgeVersion` value from the `if` conditional. 

With this, each time the `isEdgeConnected` method is triggered when an PWA connects after a disconnection from an Edge Device, the edge device's version will be fetched.

File: src/views/Settings.vue (hunk 1 above)
Comment (vickywane) [reply src/views/Settings.vue]: Can you please explain more on this. 

I seem to be lost by this comment. 
The `remoteEdgeAddress` field is not being used here neither is there a new change within this PR that has to do with it.

Comment (ivelin) [reply src/views/Settings.vue]: OK, I looked at this again and I don't see an issue with the code. Somehow I confused myself. Thanks for renaming the variable name so it's more obvious what its used for.

**Comment by ivelin** [comment]
@vickywane  This PR looks good to men now.
I see that the testsuite in the CI is failing [due to dependency error](https://github.com/ambianic/ambianic-ui/pull/722/checks?check_run_id=3534573769#step:5:16616). Is this another one related to `yarn.lock` confusing the build?

```
  > Module build failed: Error: ENOENT: no such file or directory, open '/home/runner/work/ambianic-ui/ambianic-ui/node_modules/node-libs-browser/node_modules/punycode/punycode.js'
```

**Comment by vickywane** [comment]
> I see that the testsuite in the CI is failing due to dependency error. Is this another one related to yarn.lock confusing the build?

I have fixed the lock files in the most recent commit

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane did you mean to add test cases to cover the change in behavior?

**Comment by ivelin** [comment]
> [x] 100% test coverage

@vickywane the test coverage box is checked, but I don't see tests in this PR.

**Comment by ivelin** [comment]
@vickywane I thought we agreed that two lock files is unnecessary maintenance headache. Please drop `yarn.lock`.

**Comment by ivelin** [comment]
> > I see that the testsuite in the CI is failing due to dependency error. Is this another one related to yarn.lock confusing the build?
> 
> I have fixed the lock files in the most recent commit

But you did not remove yarn.lock...

**Comment by vickywane** [comment]
@ivelin I had sufficent tests on this PR but while trying to fix the error from the DCO check previous commits were rebased and the PR was taken back to an old state. 

I had updated the PR now to include unit tests for the new behaviors within this PR. The `yarn.lock` file has also been deleted.

**Review by ivelin — COMMENTED** [review]
> @ivelin I had sufficent tests on this PR but while trying to fix the error from the DCO check previous commits were rebased and the PR was taken back to an old state.
> 
> I had updated the PR now to include unit tests for the new behaviors within this PR. The `yarn.lock` file has also been deleted.

Ouch. Be careful when merging branches.

**Review by ivelin — CHANGES_REQUESTED** [review]
We need to look more thoroughly at the state machines mismatches. See comments and recorded videos. Take your time to implement correctly.

File: src/views/Settings.vue [hunk 5]
```diff
@@ -349,53 +365,62 @@ export default {
...
           break
         default:
           break
       }
-      return step
+      return connectStep
     }
   },
+  computed: {
+    peerConnectionError: function () {
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/views/Settings.vue:398]: Alright. 

Reviewing this further, I can see that this computed property is not being  used within the component. 
Are we leaving it for further use or do we remove it?

File: src/views/Settings.vue [hunk 6]
```diff
@@ -208,6 +210,15 @@
                   >
                     <v-icon v-if="$vuetify.breakpoint.xsOnly">
                       mdi-wifi
+                      <p class="text">
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/views/Settings.vue:213]: I have gone through this several times and I think there is an underlying problem with the `pnp` module. 

The new changes within the `Settings.vue` only tap into the `peerConnectionStatus` state and are changed as the `peerConnectionStatus` events are emitted. Same way the `Navbar` uses the `peerConnectionStatus` events.

File: src/views/Settings.vue [hunk 7]
```diff
@@ -34,11 +34,12 @@
                 />
                 <amb-banner
                   v-else
-                  progress
+                  :progress="peerConnectionStatus !== 'PEER_DISCONNECTED'"
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/views/Settings.vue:37]: I have gone through this several times and I think there is an underlying problem with the `pnp` module. 

The new changes within the `Settings.vue` only tap into the `peerConnectionStatus` state and are changed as the `peerConnectionStatus` events are emitted. Same way the `Navbar` uses the `peerConnectionStatus` events.

File: src/views/Settings.vue (hunk 6 above)
[the remark this answers is not in the document]
Comment (vickywane) [reply src/views/Settings.vue:213]: This has been corrected by using an accordion to render collapsible text when user screen is small.

**Comment by ivelin** [comment]
@vickywane can you upload a video walking through the scenarios where I found problems to demonstrate they have been all resolved?

**Comment by vickywane** [comment]
> can you upload a video walking through the scenarios where I found problems to demonstrate they have been all resolved?

I have fixed the local-peer text and the `isPeerConnectionErrror` comments.

File: src/views/Settings.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (ivelin) [reply src/views/Settings.vue:37]: I see your comment but not sure what your question is. You are stating that there is an underlying problem without providing context, how to reproduce it and how to fix it. If you really think there is one, open a Bug Report issue with all the proper details.

**Comment by ivelin** [comment]
> > can you upload a video walking through the scenarios where I found problems to demonstrate they have been all resolved?
> 
> I have fixed the local-peer text and the `isPeerConnectionErrror` comments.

This does not answer my request: "can you upload a video..."



> 
> I wanted to call your attention to[ my comment](https://github.com/ambianic/ambianic-ui/pull/722#discussion_r707302047) on the out-of-sync settings page.

Responded. I did not respond earlier because it was not a clearly formulated question.

**Comment by ivelin** [comment]
@vickywane there are still unresolved conversations. Please do not submit review request until all comments are resolved.

Screen Shot 2021-09-13 at 9 03 11 PM

**Comment by vickywane** [comment]
> > > can you upload a video walking through the scenarios where I found problems to demonstrate they have been all resolved?
> > 
> > 
> > I have fixed the local-peer text and the `isPeerConnectionErrror` comments.
> 
> This does not answer my request: "can you upload a video..."

Yes i can. 

I am however being blocked being by this issue that's preventing the `ambianic-edge` docker image from connecting to my PWA for development.

**Comment by ivelin** [comment]
https://docs.ambianic.ai/developers/e2e-manual-testing/

**Comment by vickywane** [comment]
I am fixing this [issue](https://github.com/ambianic/ambianic-ui/issues/710) within this pull request as it affects this pull request. 

@ivelin I just pushed [this commit](https://github.com/vickywane/ambianic-ui/commit/d39cdf9af911ed59897a7e1c4bf4ea4e751f7016) to the existing pull request. 

Within the commit i addressed the "peer connection not getting disconnected" bug by manually disconnecting the peer instance using the [disconnect](https://peerjs.com/docs.html#peerdisconnect) method. 

This works as the `peer.on("disconnected")` event is fired, but I would like you to have a look at the code and let me know your thoughts on it.    

I will continue this PR by implementing a dialog to inform users that the existing connection is about to be disconnected.

**Comment by vickywane** [comment]
@ivelin 

I have worked further on this pull request since the last review. 

**Note:** The bug of the "PWA maintaining a background connection to an edge device" is being handled within this[ pull request](https://github.com/ambianic/ambianic-ui/pull/727), so the work done here does not cover it.

The items below are a rundown of what has been done on the pull request based on discussions on Slack;

- The "Pair with local Ambianic Edge device " text has been made adaptable to smaller resolutions by using a media query with an accordion. On a small resolution, the text becomes collapsible which the user can click to expand or collapse. On large devices, the text is fully shown to occupy the display width.  

- The edge device loading component is now only shown when the PWA is **connected** and not when **connecting** to avoid confusions.

- A dialog has been introduced to inform the user that a disconnection from the connected edge device will happen. The user can either **confirm** or **cancel** the action.

The video below shows a demonstration of the items above; 
https://user-images.githubusercontent.com/29664439/134020103-01c15e85-07b7-4774-a362-250d4d11e42c.mp4

**Comment by ivelin** [comment]
OK, I will review this. I am working on Custom Edge Device Name
<https://github.com/ambianic/ambianic-ui/issues/702> that affects the
settings page so we will need to merge code.

In the meanwhile , you can move on to a different issue. For example this
one <https://github.com/ambianic/ambianic-ui/issues/725>.

**Review by ivelin — CHANGES_REQUESTED** [review]
Please take care of DCO check.

**Comment by vickywane** [comment]
@ivelin The DCO check has been fixed.

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane please sync with the upstream master branch address the unresolved conversation.

**Review by ivelin — CHANGES_REQUESTED** [review]

**Pull request closed without merging**